File: apps/web/src/apis/me.ts
/**
 * Me API — liked Plaza items for the current user.
 */

import { request } from '@/utils/request';
import type { PlazaFeedItemDto } from '@/apis/plaza';

export type MeLikedItemDto = PlazaFeedItemDto & {
  likedAt?: number;
  coverDocument?: unknown | null;
};

export type PaginatedMeLiked = {
  items: MeLikedItemDto[];
  page: number;
  pageSize: number;
  total: number;
  hasMore: boolean;
};

export const fetchMyLiked = (params: { page: number; pageSize: number }) =>
  request<PaginatedMeLiked>({
    url: '/api/v1/me/liked',
    method: 'get',
    params,
    // Always fresh after like/unlike — do not reuse StrictMode GET cache.
    skipInflightDedupe: true,
  });

let _likedIdsCache: { at: number; data: { ids: string[] } } | null = null;
const LIKED_IDS_TTL_MS = 60_000;

export function invalidateMyLikedIdsCache() {
  _likedIdsCache = null;
}

export async function fetchMyLikedIds(): Promise<{ ids: string[] }> {
  if (_likedIdsCache && Date.now() - _likedIdsCache.at < LIKED_IDS_TTL_MS) {
    return _likedIdsCache.data;
  }
  const data = await request<{ ids: string[] }>({
    url: '/api/v1/me/liked/ids',
    method: 'get',
    skipInflightDedupe: true,
  });
  _likedIdsCache = { at: Date.now(), data };
  return data;
}

export async function likePlazaItem(submissionId: string) {
  const data = await request<{ ok: boolean; liked: boolean; id: string; likeCount?: number }>({
    url: `/api/v1/me/liked/${encodeURIComponent(submissionId)}`,
    method: 'put',
  });
  invalidateMyLikedIdsCache();
  return data;
}

export async function unlikePlazaItem(submissionId: string) {
  const data = await request<{ ok: boolean; liked: boolean; id: string; likeCount?: number }>({
    url: `/api/v1/me/liked/${encodeURIComponent(submissionId)}`,
    method: 'delete',
  });
  invalidateMyLikedIdsCache();
  return data;
}

/** Migrate legacy localStorage like ids → server. */
export async function syncMyLiked(ids: string[]) {
  const data = await request<{ ok: boolean; ids: string[] }>({
    url: '/api/v1/me/liked/sync',
    method: 'post',
    data: { ids },
  });
  invalidateMyLikedIdsCache();
  return data;
}

/** BYOK provider vault — list never includes plaintext apiKey. */
export type ByokProviderDto = {
  id: string;
  name: string;
  website?: string;
  baseUrl: string;
  apiModel?: string;
  modelKind: string;
  apiKeyHint?: string;
  hasApiKey?: boolean;
  createdAt?: number;
  updatedAt?: number;
};

export async function fetchByokProviders(): Promise<ByokProviderDto[]> {
  const data = await request<{ items: ByokProviderDto[] }>({
    url: '/api/v1/me/byok/providers',
    method: 'get',
    skipInflightDedupe: true,
  });
  return data.items || [];
}

export async function upsertByokProvider(body: {
  id?: string;
  name: string;
  website?: string;
  baseUrl: string;
  apiModel: string;
  modelKind?: string;
  apiKey?: string;
}): Promise<ByokProviderDto> {
  const data = await request<{ item: ByokProviderDto }>({
    url: '/api/v1/me/byok/providers',
    method: 'put',
    data: body,
  });
  return data.item;
}

export const deleteByokProvider = (providerId: string) =>
  request<{ ok: boolean }>({
    url: `/api/v1/me/byok/providers/${encodeURIComponent(providerId)}`,
    method: 'delete',
  });

File: apps/api/api/v1/me.py
"""Me API — liked Plaza items + BYOK provider vault for the current user."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Header, HTTPException
from pydantic import BaseModel, Field

from services.auth import get_session
from services.me import likes as likes_store
from services.security import (
    delete_byok_provider,
    list_byok_providers,
    upsert_byok_provider,
)

router = APIRouter()


def _bearer(authorization: str | None) -> str | None:
    if not authorization:
        return None
    parts = authorization.split(" ", 1)
    if len(parts) == 2 and parts[0].lower() == "bearer":
        return parts[1].strip()
    return None


def _require_user(authorization: str | None):
    user = get_session(_bearer(authorization))
    if not user:
        raise HTTPException(status_code=401, detail="Unauthorized")
    return user


class SyncLikedIn(BaseModel):
    ids: list[str] = Field(default_factory=list)


class ByokProviderIn(BaseModel):
    id: str | None = None
    name: str = ""
    website: str = ""
    baseUrl: str = ""
    apiModel: str = ""
    modelKind: str = "text"
    # Omit or empty on update to keep existing encrypted key.
    apiKey: str | None = None


@router.get("/liked")
def me_liked_list(
    page: int = 1,
    pageSize: int = 24,
    authorization: str | None = Header(default=None),
) -> dict[str, Any]:
    user = _require_user(authorization)
    return likes_store.list_liked(user.id, page=page, page_size=pageSize)


@router.get("/liked/ids")
def me_liked_ids(authorization: str | None = Header(default=None)) -> dict[str, Any]:
    user = _require_user(authorization)
    return {"ids": likes_store.list_liked_ids(user.id)}


@router.put("/liked/{submission_id}")
def me_like(
    submission_id: str,
    authorization: str | None = Header(default=None),
) -> dict[str, Any]:
    user = _require_user(authorization)
    try:
        return likes_store.like_submission(user.id, submission_id)
    except LookupError:
        raise HTTPException(status_code=404, detail="Submission not found") from None
    except ValueError as err:
        raise HTTPException(status_code=400, detail=str(err)) from err


@router.delete("/liked/{submission_id}")
def me_unlike(
    submission_id: str,
    authorization: str | None = Header(default=None),
) -> dict[str, Any]:
    user = _require_user(authorization)
    return likes_store.unlike_submission(user.id, submission_id)


@router.post("/liked/sync")
def me_liked_sync(
    body: SyncLikedIn,
    authorization: str | None = Header(default=None),
) -> dict[str, Any]:
    """One-shot migrate from client-local like ids."""
    user = _require_user(authorization)
    return likes_store.sync_likes(user.id, body.ids or [])


@router.get("/byok/providers")
def me_byok_list(authorization: str | None = Header(default=None)) -> dict[str, Any]:
    """List BYOK providers — never returns plaintext apiKey."""
    user = _require_user(authorization)
    return {"items": list_byok_providers(user.id)}


@router.put("/byok/providers")
def me_byok_upsert(
    body: ByokProviderIn,
    authorization: str | None = Header(default=None),
) -> dict[str, Any]:
    user = _require_user(authorization)
    try:
        item = upsert_byok_provider(
            user.id,
            provider_id=body.id,
            name=body.name,
            website=body.website,
            base_url=body.baseUrl,
            model_kind=body.modelKind,
            api_key=body.apiKey,
            api_model=body.apiModel,
        )
    except ValueError as err:
        raise HTTPException(status_code=400, detail=str(err)) from err
    return {"item": item}


@router.delete("/byok/providers/{provider_id}")
def me_byok_delete(
    provider_id: str,
    authorization: str | None = Header(default=None),
) -> dict[str, Any]:
    user = _require_user(authorization)
    ok = delete_byok_provider(user.id, provider_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Provider not found")
    return {"ok": True}
